Build failure: python3Packages.docling-parse

### Nixpkgs version

- Unstable (25.11)

### Steps to reproduce

Switch to branch `nixos-unstable-small`, and run:

```
nix build .#python3Packages.docling-parse
```

### Can Hydra reproduce this build failure?

Hydra doesn’t try to build the package.

### Link to Hydra build job

_No response_

### Relevant log output

```console
Executing pytestCheckPhase
pytest flags: -m pytest
============================= test session starts ==============================
platform linux -- Python 3.13.5, pytest-8.3.5, pluggy-1.6.0
rootdir: /build/source
configfile: pyproject.toml
collected 14 items                                                             

tests/test_parse.py F...                                                 [ 28%]
tests/test_parse_v1.py ....                                              [ 57%]
tests/test_parse_v2.py ......                                            [100%]

=================================== FAILURES ===================================
___________________ test_reference_documents_from_filenames ____________________

    def test_reference_documents_from_filenames():
    
        parser = DoclingPdfParser(loglevel="fatal")
    
        pdf_docs = sorted(glob.glob(REGRESSION_FOLDER))
    
        assert len(pdf_docs) > 0, "len(pdf_docs)==0 -> nothing to test"
    
        for pdf_doc_path in pdf_docs:
            print(f"parsing {pdf_doc_path}")
    
            pdf_doc: PdfDocument = parser.load(
                path_or_stream=pdf_doc_path,
                boundary_type=PdfPageBoundaryType.CROP_BOX,  # default: CROP_BOX
                lazy=False,
            )  # default: True
            assert pdf_doc is not None
    
            # PdfDocument.iterate_pages() will automatically populate pages as they are yielded.
            # No need to call PdfDocument.load_all_pages() before.
            for page_no, pred_page in pdf_doc.iterate_pages():
                # print(f" -> Page {page_no} has {len(pred_page.sanitized.cells)} cells.")
    
                rname = os.path.basename(pdf_doc_path)
                fname = os.path.join(
                    GROUNDTRUTH_FOLDER, rname + f".page_no_{page_no}.py.json"
                )
    
                if GENERATE or (not os.path.exists(fname)):
                    pred_page.save_as_json(fname)
    
                    for unit in [TextCellUnit.CHAR, TextCellUnit.WORD, TextCellUnit.LINE]:
                        lines = pred_page.export_to_textlines(
                            cell_unit=unit,
                            add_fontkey=True,
                            add_fontname=False,
                        )
                        _fname = fname + f".{unit}.txt"
                        with open(_fname, "w") as fw:
                            fw.write("\n".join(lines))
                else:
                    # print(f"loading from {fname}")
    
                    for unit in [TextCellUnit.CHAR, TextCellUnit.WORD, TextCellUnit.LINE]:
                        _lines = pred_page.export_to_textlines(
                            cell_unit=unit,
                            add_fontkey=True,
                            add_fontname=False,
                        )
    
                        _fname = fname + f".{unit}.txt"
                        with open(_fname, "r") as fr:
                            lines = fr.readlines()
    
                        olines = "".join(lines)
                        _olines = "\n".join(_lines)
    
>                       assert olines == _olines, "olines==_olines"
E                       AssertionError: olines==_olines
E                       assert '(559.67, 010...      /C0_0 は' == '(559.67, 010...FT_TO_RIGHT は'
E                         
E                         (pytest_assertion plugin: representation of details failed: /nix/store/djck7mx6jad1w0yy6zings96dyxanls6-python3-3.13.5/lib/python3.13/difflib.py:193: RecursionError: maximum recursion depth exceeded.
E                          Probably an object has a faulty __repr__.)

tests/test_parse.py:266: AssertionError
----------------------------- Captured stdout call -----------------------------
parsing tests/data/regression/cropbox_versus_mediabox_01.pdf
=========================== short test summary info ============================
FAILED tests/test_parse.py::test_reference_documents_from_filenames - AssertionError: olines==_olines
=================== 1 failed, 13 passed in 220.20s (0:03:40) ===================
```

### Additional context

This seems to be introduced by applying both #425736 and #424693:

* If Docling-parse stays at 4.0.5, upgrading docling-core to latest version 2.43.0 can pass all tests.
* If Docling-parse is bumped to 4.1.0, the last known working docling-core version is 2.38.1 (I manually bisected it).


I checked the version constraints of all docling related packages, seems downgrading docling-core to 2.38.1 still satisfy all conditions. Not sure if downgrading docling-core to mitigate the build issue follow the recommended practice in maintaining Nixpkgs. If so, I can do this with an upcoming PR.


### System metadata

 - system: `"x86_64-linux"`
 - host os: `Linux 6.12.37, NixOS, 25.05 (Warbler), 25.05.20250718.77b13cb`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.28.4`
 - nixpkgs: `/nix/store/vqzkgbcl706x2ib311a95xvaynvklsyi-source`

### Notify maintainers

@drupol 

---

**Note for maintainers:** Please tag this issue in your pull request description. (i.e. `Resolves #ISSUE`.)


### I assert that this issue is relevant for Nixpkgs

- [x] I assert that this is a bug and not a support request.
- [x] I assert that this is not a [duplicate of an existing issue](https://github.com/NixOS/nixpkgs/issues?q=is%3Aopen+is%3Aissue+label%3A%220.kind%3A+build+failure%22). 
- [x] I assert that I have read the [NixOS Code of Conduct](https://github.com/NixOS/.github/blob/master/CODE_OF_CONDUCT.md) and agree to abide by it.

### Is this issue important to you?

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

--- a/pkgs/development/python-modules/docling-core/default.nix
+++ b/pkgs/development/python-modules/docling-core/default.nix
@@ -29,14 +29,14 @@
 
 buildPythonPackage rec {
   pname = "docling-core";
-  version = "2.43.0";
+  version = "2.38.1";
   pyproject = true;
 
   src = fetchFromGitHub {
     owner = "docling-project";
     repo = "docling-core";
     tag = "v${version}";
-    hash = "sha256-c9TaX4INfTfR3ZpmXbOteHr2R2jAbVzvMk8tO1XV4Nc=";
+    hash = "sha256-zfpinXjZrGp5SF5IXLSP/A9Kb1WzqXXtIhhk/G3nY4k=";
   };
 
   build-system = [

--- a/pkgs/development/python-modules/docling-jobkit/default.nix
+++ b/pkgs/development/python-modules/docling-jobkit/default.nix
@@ -1,0 +1,108 @@
+{
+  lib,
+  buildPythonPackage,
+  fetchFromGitHub,
+
+  # build-system
+  hatchling,
+  poetry-core,
+
+  # dependencies
+  docling,
+  pydantic-settings,
+  typer,
+  boto3,
+  pandas,
+  fastparquet,
+  pyarrow,
+  httpx,
+
+  # optional dependencies
+  tesserocr,
+  rapidocr-onnxruntime,
+  onnxruntime,
+  ray,
+
+  # tests
+  pytestCheckHook,
+  pytest-asyncio,
+  writableTmpDirAsHomeHook,
+
+  # options
+  withTesserocr ? false,
+  withRapidocr ? false,
+  withRay ? false,
+}:
+
+buildPythonPackage rec {
+  pname = "docling-jobkit";
+  version = "1.1.0";
+  pyproject = true;
+
+  src = fetchFromGitHub {
+    owner = "docling-project";
+    repo = "docling-jobkit";
+    tag = "v${version}";
+    hash = "sha256-V+Sc/eEuv0Q3q/wjGpkraD6ZR4Kfhorc3meg7fy0LTA=";
+  };
+
+  build-system = [
+    hatchling
+    poetry-core
+  ];
+
+  dependencies =
+    [
+      docling
+      pydantic-settings
+      typer
+      boto3
+      pandas
+      fastparquet
+      pyarrow
+      httpx
+    ]
+    ++ lib.optionals withTesserocr optional-dependencies.tesserocr
+    ++ lib.optionals withRapidocr optional-dependencies.rapidocr
+    ++ lib.optionals withRay optional-dependencies.ray;
+
+  optional-dependencies = {
+    tesserocr = [ tesserocr ];
+    rapidocr = [
+      rapidocr-onnxruntime
+      onnxruntime
+    ];
+    ray = [ ray ];
+  };
+
+  pythonRelaxDeps = [
+    "boto3"
+    "pyarrow"
+  ];
+
+  pythonImportsCheck = [
+    "docling"
+    "docling_jobkit"
+  ];
+
+  nativeCheckInputs = [
+    pytestCheckHook
+    pytest-asyncio
+    writableTmpDirAsHomeHook
+  ];
+
+  disabledTests = [
+    # requires network access
+    "test_convert_url"
+    "test_convert_file"
+    "test_convert_warmup"
+  ];
+
+  meta = {
+    changelog = "https://github.com/docling-project/docling-jobkit/blob/${src.tag}/CHANGELOG.md";
+    description = "Running a distributed job processing documents with Docling";
+    homepage = "https://github.com/docling-project/docling-jobkit";
+    license = lib.licenses.mit;
+    maintainers = with lib.maintainers; [ codgician ];
+  };
+}

--- a/pkgs/development/python-modules/docling-serve/default.nix
+++ b/pkgs/development/python-modules/docling-serve/default.nix
@@ -6,10 +6,12 @@
   setuptools-scm,
   # python dependencies
   docling,
+  docling-jobkit,
   fastapi,
   httpx,
   pydantic-settings,
   python-multipart,
+  scalar-fastapi,
   uvicorn,
   websockets,
   tesserocr,
@@ -29,20 +31,15 @@
 
 buildPythonPackage rec {
   pname = "docling-serve";
-  version = "0.14.0";
+  version = "1.0.0";
   pyproject = true;
 
   src = fetchFromGitHub {
     owner = "docling-project";
     repo = "docling-serve";
     tag = "v${version}";
-    hash = "sha256-R8W/FXKj2wLJOcjwIsna/2wFOLGM80Qr3WlYPJTTSNU=";
+    hash = "sha256-kvtsutdQXk9mxx90fHXlzxFXrA7v+TdlgjrKCSFuuVU=";
   };
-
-  postPatch = ''
-    substituteInPlace pyproject.toml \
-      --replace-fail '"kfp[kubernetes]>=2.10.0",' ""
-  '';
 
   build-system = [
     hatchling
@@ -60,10 +57,12 @@ buildPythonPackage rec {
   dependencies =
     [
       docling
+      (docling-jobkit.override { inherit withTesserocr withRapidocr; })
       fastapi
       httpx
       pydantic-settings
       python-multipart
+      scalar-fastapi
       typer
       uvicorn
       websockets

--- a/pkgs/development/python-modules/docling/default.nix
+++ b/pkgs/development/python-modules/docling/default.nix
@@ -52,14 +52,14 @@
 
 buildPythonPackage rec {
   pname = "docling";
-  version = "2.41.0";
+  version = "2.38.1";
   pyproject = true;
 
   src = fetchFromGitHub {
     owner = "docling-project";
     repo = "docling";
     tag = "v${version}";
-    hash = "sha256-GD052HCqBLs+KUkOUOVdlXxS6+PD2pthGtz+zdQ6QnM=";
+    hash = "sha256-ITRpWO7PEtfg5kM9LF3koZgzuPtLxoGFhorMlXlwSdI=";
   };
 
   build-system = [

--- a/pkgs/development/python-modules/scalar-fastapi/default.nix
+++ b/pkgs/development/python-modules/scalar-fastapi/default.nix
@@ -1,0 +1,65 @@
+{
+  lib,
+  buildPythonPackage,
+  fetchPypi,
+
+  # build-system
+  setuptools,
+
+  # python dependencies
+  annotated-types,
+  anyio,
+  fastapi,
+  idna,
+  pydantic,
+  sniffio,
+  starlette,
+  typing-extensions,
+
+  # tests
+  pytestCheckHook,
+  httpx,
+}:
+
+buildPythonPackage rec {
+  pname = "scalar-fastapi";
+  version = "1.2.2";
+  pyproject = true;
+
+  src = fetchPypi {
+    pname = "scalar_fastapi";
+    inherit version;
+    hash = "sha256-r5GoX6f7Ru078WcRvGNyhyYDxL1w4yjbfHvlf+ZF/sc=";
+  };
+
+  build-system = [
+    setuptools
+  ];
+
+  dependencies = [
+    annotated-types
+    anyio
+    fastapi
+    idna
+    pydantic
+    sniffio
+    starlette
+    typing-extensions
+  ];
+
+  pythonImportsCheck = [
+    "scalar_fastapi"
+  ];
+
+  nativeCheckInputs = [
+    pytestCheckHook
+    httpx
+  ];
+
+  meta = {
+    description = "A plugin for FastAPI to render a reference for your OpenAPI document";
+    homepage = "https://github.com/scalar/scalar/";
+    license = lib.licenses.mit;
+    maintainers = with lib.maintainers; [ codgician ];
+  };
+}

--- a/pkgs/top-level/python-packages.nix
+++ b/pkgs/top-level/python-packages.nix
@@ -15951,6 +15951,8 @@ self: super: with self; {
 
   sbom4files = callPackage ../development/python-modules/sbom4files { };
 
+  scalar-fastapi = callPackage ../development/python-modules/scalar-fastapi { };
+
   scalene = callPackage ../development/python-modules/scalene { };
 
   scales = callPackage ../development/python-modules/scales { };

--- a/pkgs/top-level/python-packages.nix
+++ b/pkgs/top-level/python-packages.nix
@@ -4192,6 +4192,8 @@ self: super: with self; {
 
   docling-ibm-models = callPackage ../development/python-modules/docling-ibm-models { };
 
+  docling-jobkit = callPackage ../development/python-modules/docling-jobkit { };
+
   docling-parse = callPackage ../development/python-modules/docling-parse {
     loguru-cpp = pkgs.loguru;
   };

